packages/@react-stately/list/src/useListState.ts

```diff
@@ -88,48 +88,68 @@ function useFocusedKeyReset<T>(collection: Collection<Node<T>>, selectionManager
   useEffect(() => {
-    if (selectionManager.focusedKey != null && !collection.getItem(selectionManager.focusedKey) && cachedCollection.current) {
-      const startItem = cachedCollection.current.getItem(selectionManager.focusedKey);
-      const cachedItemNodes = [...cachedCollection.current.getKeys()].map(
-        key => {
-          const itemNode = cachedCollection.current!.getItem(key);
-          return itemNode?.type === 'item' ? itemNode : null;
-        }
-      ).filter(node => node !== null);
-      const itemNodes = [...collection.getKeys()].map(
-        key => {
-          const itemNode = collection.getItem(key);
-          return itemNode?.type === 'item' ? itemNode : null;
-        }
-      ).filter(node => node !== null);
-      const diff: number = (cachedItemNodes?.length ?? 0) - (itemNodes?.length ?? 0);
-      let index = Math.min(
-        (
-          diff > 1 ?
-          Math.max((startItem?.index ?? 0) - diff + 1, 0) :
-          startItem?.index ?? 0
-        ),
-        (itemNodes?.length ?? 0) - 1);
-      let newNode: Node<T> | null = null;
-      let isReverseSearching = false;
-      while (index >= 0) {
-        if (!selectionManager.isDisabled(itemNodes[index].key)) {
-          newNode = itemNodes[index];
-          break;
+    if (selectionManager.focusedKey != null && cachedCollection.current) {
+      let item = collection.getItem(selectionManager.focusedKey);
+      if (!item || (item.type === 'loader' && !item.props.isLoading)) {
+        const startItem = cachedCollection.current.getItem(selectionManager.focusedKey);
+        const cachedItemNodes = [...cachedCollection.current.getKeys()].map(
+          key => {
+            const itemNode = cachedCollection.current!.getItem(key);
+            return itemNode?.type === 'item' || itemNode?.type === 'loader' ? itemNode : null;
+          }
+        ).filter(node => node !== null);
+        const itemNodes = [...collection.getKeys()].map(
+          key => {
+            const itemNode = collection.getItem(key);
+            return itemNode?.type === 'item' || itemNode?.type === 'loader' ? itemNode : null;
+          }
+        ).filter(node => node !== null);
+        const diff: number = (cachedItemNodes?.length ?? 0) - (itemNodes?.length ?? 0);
+        // Look up the start item's index in the cached node array. We can't rely on the node's index
+        // because it maybe in a section and thus have a index relative to that section
+        // TODO: the collection keys/item are not guarenteed to be in order
+        // (aka the loader will be the first node in the keymap if the collection was empty and then loaded items)
+        // this means we should ideally traverse through the key's prevKey until we find a focusable node but
+        // that is a bigger refactor, will handle later. For now continue assuming that the loader is the last element
+        let startItemIndex = 0;
+        if (startItem) {
```

Comment:
To avoid further bloat, will be handling this refactor later. Perhaps something can be done in the RAC collection level but otherwise we should maybe consider refactoring this to use the old focused key's `prevKey` instead of index since the collection's keymap order is not necessarily in index order. 

Will need to discuss where exactly we want to restore focus (in roughly the same place? or should it prefer the closest contextually matching row regardless of where that maybe in the collection?)